src/librbd/ImageWatcher.cc

```diff
@@ -590,6 +590,7 @@ void ImageWatcher::acknowledge_notify(uint64_t notify_id, uint64_t handle,
 
 void ImageWatcher::reregister_watch() {
   ldout(m_image_ctx.cct, 10) << "re-registering image watch" << dendl;
+  handle_header_update();
```

Comment:
I think this should be after the watch is re-established so we don't miss any header updates between getting the error and re-registering. Once the watch is in place we'll get notifies again. Taking owner_lock exclusively like it does already prevents new writes from proceeding in the meantime.